django/db/models/query.py

```diff
@@ -890,31 +890,73 @@ def bulk_update(self, objs, fields, batch_size=None):
                     attr = getattr(obj, field.attname)
-                    if not hasattr(attr, "resolve_expression"):
+                    if hasattr(attr, "resolve_expression"):
+                        has_field_references = True
+                    else:
                         attr = Value(attr, output_field=field)
-                    when_statements.append(When(pk=obj.pk, then=attr))
-                case_statement = Case(*when_statements, output_field=field)
-                if requires_casting:
-                    case_statement = Cast(case_statement, output_field=field)
-                update_kwargs[field.attname] = case_statement
-            updates.append(([obj.pk for obj in batch_objs], update_kwargs))
+
+                    if requires_casting_in_values:
+                        attr = Cast(attr, output_field=field)
+                    values.append(attr)
+
+                row_tuples.append(RowTuple(*values))
+            updates.append(
+                RowTupleValues(
+                    row_tuples, pk_field=self.model._meta.pk, field_list=fields
+                )
+            )
+
         rows_updated = 0
-        queryset = self.using(self.db)
         with transaction.atomic(using=self.db, savepoint=False):
-            for pks, update_kwargs in updates:
-                rows_updated += queryset.filter(pk__in=pks).update(**update_kwargs)
+            for row_tuples in updates:
+                if has_field_references or has_related_fields:
+                    rows_updated += self._bulk_update_slow(
+                        row_tuples, requires_casting=requires_casting
+                    )
+                else:
+                    query = sql.BulkUpdateQuery(self.model, row_tuples=row_tuples)
+                    rows_updated += query.get_compiler(using=self.db).execute_sql(
+                        result_type=CURSOR
+                    )
         return rows_updated
 
     bulk_update.alters_data = True
 
+    def _bulk_update_slow(self, row_tuples, requires_casting=False):
+        pks = [
+            row_tuple.source_expressions[0].value
+            for row_tuple in row_tuples.source_expressions
+        ]
+        update_kwargs = {}
+        # Skip the ID column
+        for field_idx, field in enumerate(row_tuples.field_list, start=1):
+            when_statements = []
+            for row_tuple in row_tuples.source_expressions:
+                row_tuple: RowTuple
+                id_expression = row_tuple.source_expressions[0]
+                attr_expression = row_tuple.source_expressions[field_idx]
+                when_statements.append(When(pk=id_expression, then=attr_expression))
+            case_statement = Case(*when_statements, output_field=field)
+            if requires_casting:
+                case_statement = Cast(case_statement, output_field=field)
+            update_kwargs[field.attname] = case_statement
+
+        return self.filter(pk__in=pks).update(**update_kwargs)
```

Comment:
I wonder if it might make sense to push that down to the compiler itself that could use _raw_ SQL strings directly over slow expression resolving.